k4FWCore/components/PodioInput.cpp

```diff
@@ -184,14 +184,15 @@ PodioInput::PodioInput(const std::string& name, ISvcLocator* svcLoc) : Consumer(
     error() << "Could not get PodioDataSvc" << endmsg;
   }
   fillReaders();
-
-  auto key = edm4hep::EventHeaderName;
-  if (std::find(m_collectionNames.begin(), m_collectionNames.end(), key) == m_collectionNames.end()) {
-    m_collectionNames.value().push_back(key);
-  }
 }
 
 void PodioInput::operator()() const {
+  if (m_podioDataSvc->hasCollection(edm4hep::EventHeaderName)) {
+    m_readers[edm4hep::EventHeaderCollection::typeName](edm4hep::EventHeaderName);
```

Comment:
I thought about doing it this way initially, what happens if the EventHeader collection is also in the list of collections to read? Will it be read twice?

Reply:
It would query the underlying Frame again for a second time. However, the second time around that effectively is a map lookup and return. Technically, `hasCollection` does all the "heavy lifting" of prompting the Frame to actually read the data and populate it's internal map. Most importantly you will always get the same collection.

I quickly checked what happens if I explicitly make the test read the EventHeader collection and nothing seems to have broken, but it looks like we would be putting two DataHandles pointing to the same collection into the EDS at the moment. I have to check why that seems to work without breaking.

Reply:
We indeed do get two `DataHandles` (and corresponding objects in the EDS) if `EventHeader` is explicitly passed in. However, since the collection is owned by the Frame and the DataSvc defers the cleanup to that, there is at least no double free.

I will add a simple filter to clean out duplicate collection names as well as `"EventHeader"`. Additionally, I will adapt the `PodioDataSvc::readCollection` to check for existence in the EDS before populating it twice.

Reply:
Didn't do the filtering in the end, because it is effectively unnecessary with the guard against duplicate entries in the EDS.